### ISSUE TITLE
Remove precompiler flags from CHMM

### ISSUE THREAD

**Issue by karlnapf** [shogun-toolbox/shogun#2712]
Both from CMakeLists.txt and the code it self. The corresponding options should be part of the class interface itself, or solved via templating

**Comment by srgnuclear** [comment]
hi i would like to take up this issue and subsequently i could complete other issues related to hmm which are listed along with it. 
I went through CHMM class in hmm.h. By precompiler flags do u mean inline functions or something else? I would like to know more as to how should i proceed with it.

**Comment by iglesias** [comment]
I think it is referring to ifdef and other instructions to be taken care
before compilation.

**Comment by srgnuclear** [comment]
kk. I went through CHMM class and i found the following ifdefs:

#ifdef USE_HMMPARALLEL_STRUCTURES
#ifdef USE_LOGSUMARRAY
#ifdef FIX_POS
#ifdef USE_HMMDEBUG
#ifdef HMM_DEBUG

These are present in config.h.in and cmakelists.txt. I want to know the flow process of how things work. 
Like in cmakelists the above flags are either set of or on in option function : so here is the flag is set to of then why is it defined as 1 in cmakedefine in config.h.in like here i am confused about the correlation between the two things. 
Then an if checks if the flag is set to 1 or not : so here is the value obtained from cmakedefine in config.h.in?
Further if the  value is true it is appended to a list : so here i want to know the context of the list.

About the task : 
1) Can u provide me with a reference where options have been made part of the interface or may be solved by templating?

**Comment by iglesias** [comment]
Easy solution would be to add attributes with setters.

**Comment by karlnapf** [comment]
First step please: Basic unit testing. There are a lot of tests in the static interfaces. Those have to be translated to unit tests before we touch the class, otherwise we dont know what breaks

**Comment by srgnuclear** [comment]
ohk so i went through static interfaces in src/interfaces so there are various functions implemented in suppose RInterface.cpp so do we have to translate all of them to each unit test and for all the interfaces?

**Comment by iglesias** [comment]
@karlnapf, what are exactly the tests you mentioned in the static interfaces? I think you mean the examples in shogun/examples/undocumented/*_static should be translated to unit tests (maybe not one-to-one translation, though).

**Comment by srgnuclear** [comment]
i was going through unit tests made till now for distribution folder and there is one for mixturemodel train method. I wanted to know that will all methods of CHMM class be tested by unit tests?. For now i am beginning by train method in CHMM class.

**Comment by iglesias** [comment]
It should be like that, yes. Have a look at the other classes that are
properly unit tested to get an idea.

**Comment by srgnuclear** [comment]
i went through the examples folder and in it used java_modular implementation, which i den converted to .cc although complete unit test is yet to be implemented i want some review on the approach.
code gist: https://gist.github.com/srgnuclear/f57dfb5f59f518645952
I have commented the load part as i want to know should i be using fstream for it or is there some defined method in shogun?

**Comment by iglesias** [comment]
Unit tests should use simple data so that they are trivial. So no need to
load data files. See how it is done in other unit tests.

**Comment by karlnapf** [comment]
Yes @iglesias these static examples are the ones I mean. Many of them work on toy data so can easily be reproduced as unit tests. Certainly, all methods in the static tests should be unit tested. Since we *have* static tests, we can easily write toy examples for unit testing the same methods, using Shogun's current own outputas ground truth (since existing tests somehow guarantee that results are correct)

**Comment by srgnuclear** [comment]
i think this may be a bug in MixtureModel_unittest.cc where a semicolon is missing at the end SG_UNREF(mix) .
Like does the test run in this case ?

**Comment by iglesias** [comment]
You should try running it locally and check it
